src/gf.c

```diff
@@ -1438,7 +1445,11 @@ jl_tupletype_t *arg_type_tuple(jl_value_t **args, size_t nargs)
             else
                 types[i] = jl_typeof(ai);
         }
-        tt = (jl_tupletype_t*)jl_inst_concrete_tupletype_v(types, nargs);
+        // if `ai` has free type vars this will not be a leaf type.
+        // TODO: it would be really nice to only dispatch and cache those as
+        // `jl_typeof(ai)`, but that will require some redesign of the caching
+        // logic.
```

Comment:
I think we could handle this case if we make sure to add a guardsig with `Type` in this slot. I think the main complication would then be in detecting this case. Currently, the `cache_method` logic assume that we can always choose to avoid needing a `guardsig` by using the original `tt` as the `simplesig`, while in this case, we might need both a `simplesig` and a `guardsig` (with each being triggered by the despecialization of a different arguments).